packages/medusa/src/strategies/product-export.ts

```diff
@@ -0,0 +1,493 @@
+  protected readonly productService_: ProductService
+  protected readonly productRepository_: typeof ProductRepository
+
+  protected readonly relations_ = [...defaultAdminProductRelations]
+  protected readonly columnDescriptors = new Map<string, Accessor>([
+    ["Product Handle", (line: Product): string => line.handle],
+    ["Product Title", (line: Product): string => line.title],
+    ["Product Subtitle", (line: Product): string => line.subtitle],
+    ["Product Description", (line: Product): string => line.description],
+    ["Product Status", (line: Product): string => line.status],
+    ["Product Thumbnail", (line: Product): string => line.thumbnail],
+    ["Product Weight", (line: Product): string => line.weight?.toString()],
+    ["Product Length", (line: Product): string => line.length?.toString()],
+    ["Product Width", (line: Product): string => line.width?.toString()],
+    ["Product Height", (line: Product): string => line.height?.toString()],
+    ["Product HS Code", (line: Product): string => line.hs_code?.toString()],
+    [
+      "Product Origin Country",
+      (line: Product): string => line.origin_country?.toString(),
+    ],
+    ["Product Mid Code", (line: Product): string => line.mid_code?.toString()],
+    ["Product Material", (line: Product): string => line.material?.toString()],
+    [
+      "Product Collection Title",
+      (line: Product): string => line.collection.title,
+    ],
+    [
+      "Product Collection Handle",
+      (line: Product): string => line.collection.handle,
+    ],
+    ["Product Type", (line: Product): string => line.type.value],
+    [
+      "Product Tags",
+      (line: Product): string =>
+        (line.tags.map((t) => t.value) ?? []).join(","),
+    ],
+    [
+      "Product Discountable",
+      (line: Product): string => line.discountable?.toString(),
+    ],
+    ["Product External ID", (line: Product): string => line.external_id],
+    ["Product Profile Name", (line: Product): string => line?.profile?.name],
+    ["Product Profile Type", (line: Product): string => line?.profile?.type],
+    /*
+     *
+     * The other columns are dynamically computed via the `buildHeader` method.
+     * This will ensure that all the expected columns based on the
+     * real data and therefore the highest number of value for the different
+     * relation that we are expecting to appear
+     *
+     */
+  ])
+
+  private readonly NEWLINE = "\r\n"
+  private readonly DELIMITED_ = ";"
+  private readonly BATCH_SIZE = 50
+
+  constructor(protected readonly container: InjectedDependencies) {
+    super(container)
+
+    this.manager_ = container.manager
+    this.batchJobService_ = container.batchJobService
+    this.productService_ = container.productService
+    this.productRepository_ = container.productRepository
+
+    // Defer resolution of the fileService to avoid cycling issues with services resolving strategies and vice verca
+    return new Proxy(this, {
```

Comment:
comment: Just for the explanation 

Awilix have a resolution system that make the resolution happening only when a property from the container is requested, here `container.fileService` will call the get from the awilix underlaying proxy and therefor call the resolution at that moment.

In order to avoid cycling dependencies issues (since the strategy require the fileService, but some services need the strategies, which means the loading order can't be solved like that) I did add a proxy that defer the call of the container property `fileService` and therefor it also defer the resolution of the fileService to the moment we need it for the first time.

that way, all the components can be properly registered and the cradle is passed, but the resolution only happen when needed.
I know that in that case we are using the container as a service locator, which is the anti pattern of the dependency injection. But the fact is that everywhere in medusa it is already used as a service locator through out the destructuring that only hide the service locator usage.

@srindom @olivermrbl @pKorsholm what is your opinion on that, should I move that somewhere else, should I do it differently, I would like to get your feedback on that

Reply:
I could maybe (😂 ) find a way to make it the default behaviour over the container to avoid cyclic dependencies anywhere. This is also why it is important when using DI to not access any dependencies in the constructor as they could be undefined until proper resolution. This is a common behaviour in the DI world

Reply:
After discussing with @pKorsholm we will split the strategy loading so that the batches are loaded after the services. It will also solve the problem as long as the services does not require the batchJobStrategy in some way